### PR TITLE
add pkcs11 support to Ansible ssh connection module

### PR DESCRIPTION
**Pull request by thedavidwhiteside**
##### SUMMARY
Ansible does not currently support PKCS#11 (PIV, CAC, smartcard) authentication.  With a simple modification to the ssh connection plugin, you can easily add support for PKCS#11.  Paramiko is in the process of adding PKCS11 support, after the support is added in the next stable release of Paramiko then its also possible to add PKCS11 support for the Ansible+Paramiko combination.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
ssh connection plugin

##### ANSIBLE VERSION

```
ansible 2.11.0.dev0
  config file = None
  configured module search path = ['/Users/dwhitesi/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.9/site-packages/ansible_core-2.11.0.dev0-py3.9.egg/ansible
  ansible collection location = /Users/dwhitesi/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/local/bin/ansible
  python version = 3.9.2 (default, Feb 19 2021, 17:09:53) [Clang 12.0.0 (clang-1200.0.32.29)]
  jinja version = 3.0.0a1
  libyaml = True
```

##### USAGE
```
$ export ANSIBLE_PKCS11_PROVIDER=/usr/local/lib/opensc-pkcs11.so
$ ansible-playbook -u USERNAME -b -k -K PLAYBOOK.yml --connection=ssh
SSH password: << Enter your PKCS11 Pin for your smartcard
SUDO password[defaults to SSH password]: << Enter your user account password for sudo
```

I use https://github.com/OpenSC/OpenSC for the middleware, this is the pkcs11-provider library that you point Ansible to with the pkcs11_provider ini setting or ANSIBLE_PKCS11_PROVIDER env variable.

I tested with a smartcard, a very common device you might use with Ansible pkcs11 would be a Yubikey device (https://www.yubico.com/product/yubikey-neo).

- David

### CODE REVIEW

**Comment by bcoca** [comment]
so instead of hardocding them in many places, you can just add it instead just to the connection plugin that uses them, as ini/env/connection variables they will also be a lot more flexible.
https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/connection/ssh
.py#L131

**Comment by thedavidwhiteside** [comment]
The pkcs11_pin is sensitive and I wanted to follow the same process Ansible currently uses for passwords.  I could just reuse the password variable for the pin and use ini/env/connection variables for the pkcs11_provider, and whenever you set the pkcs11_provider then it uses PKCS11 rather than password auth.

  I see pkcs11 making its way to other connection plugins (paramiko_ssh), so thats why I initially took the approach of the variables being CLI options.

- David

**Comment by bcoca** [comment]
i think reusing password is fine and keeping the provider as separate parameter

**Comment by starboarder2001** [comment]
Just updated, tested, and verified its working with just using env/ini/connection variables.

File: lib/ansible/constants.py [hunk 1]
```diff
@@ -135,6 +135,7 @@ def set_constant(name, value, export=vars()):
     scp_extra_args=('ansible_scp_extra_args', ),
     ssh_extra_args=('ansible_ssh_extra_args', ),
     ssh_transfer_method=('ansible_ssh_transfer_method', ),
+    pkcs11_provider=('ansible_pkcs11_provider', ),
```
Comment (bcoca) [inline lib/ansible/constants.py]: again, you dont need this, just the part in the connection plugin

File: lib/ansible/config/base.yml [hunk 2]
```diff
@@ -913,6 +913,13 @@ DEFAULT_SCP_IF_SSH:
   env: [{name: ANSIBLE_SCP_IF_SSH}]
   ini:
   - {key: scp_if_ssh, section: ssh_connection}
+pkcs11_provider:
+  default: ""
+  description:
+    - "PKCS11 SmartCard provider such as opensc, example: /usr/local/lib/opensc-pkcs11.so"
+  env: [{name: ANSIBLE_PKCS11_PROVIDER}]
+  ini:
+    - {key: pkcs11_provider, section: ssh_connection}
```
Comment (bcoca) [inline lib/ansible/config/base.yml]: You don't need this, just the part in the connection plugin

File: lib/ansible/plugins/connection/ssh.py [hunk 3]
```diff
@@ -396,13 +404,25 @@ def _build_command(self, binary, *other_args):
 
         # If we want to use password authentication, we have to set up a pipe to
         # write the password to sshpass.
-
-        if self._play_context.password:
+        pkcs11_provider = C.pkcs11_provider
```
Comment (bcoca) [inline lib/ansible/plugins/connection/ssh.py]: this is incorrect, use `self.get_option`

**Review by bcoca — CHANGES_REQUESTED** [review]

**Comment by mattclay** [comment]
CI failure in unit tests: https://app.shippable.com/github/ansible/ansible/runs/57309/3/tests

**Comment by thedavidwhiteside** [comment]
I fixed the ssh unit test that I broke, had to mock get_option.

When using export ANSIBLE_PKCS11_PROVIDER=/usr/local/lib/opensc-pkcs11.so it works fine, but when I set pkcs11_provider = /usr/local/lib/opensc-pkcs11.so in ansible.cfg it fails with the below error. So basically the env variables work great, but the ini variable seems to have a \r\n appended, which I would expect Ansible get_option to strip out automatically.

fatal: [HOSTNAME]: UNREACHABLE! => {"changed": false, "msg": "Failed to connect to the host via ssh: command-line line 0: garbage at end of line; \"/usr/local/lib/pkcs11/opensc-pkcs11.so\".\r\n", "unreachable": true}

File: lib/ansible/constants.py (hunk 1 above)
Comment (thedavidwhiteside) [reply lib/ansible/constants.py]: Thanks! I was using ansible_ssh_extra_args as my example, and wasn't completely clear if this was needed or not.

File: lib/ansible/config/base.yml (hunk 2 above)
Comment (thedavidwhiteside) [reply lib/ansible/config/base.yml]: I removed this

File: lib/ansible/plugins/connection/ssh.py (hunk 3 above)
Comment (thedavidwhiteside) [reply lib/ansible/plugins/connection/ssh.py]: I switched to get_option. It wasn't exactly clear how to get at the env/ini variables, the majority of the others seemed to be in the constants object.

Comment (bcoca) [reply lib/ansible/plugins/connection/ssh.py]: since 2.4 that is not needed anymore, ssh is a bit overloaded for other plugins so some got left 'as is', but for new options, all you need to do is define them in the plugin's docs.

File: lib/ansible/plugins/connection/ssh.py [hunk 4]
```diff
@@ -396,13 +404,25 @@ def _build_command(self, binary, *other_args):
...
         # write the password to sshpass.
-
-        if self._play_context.password:
+        pkcs11_provider = self.get_option("pkcs11_provider")
+        if self._play_context.password or len(pkcs11_provider) >= 1:
             if not self._sshpass_available():
-                raise AnsibleError("to use the 'ssh' connection type with passwords, you must install the sshpass program")
+                raise AnsibleError("to use the 'ssh' connection type with passwords or pkcs11_provider, you must install the sshpass program")
 
+        if self._play_context.password and len(pkcs11_provider) >= 1:
```
Comment (bcoca) [inline lib/ansible/plugins/connection/ssh.py]: just add teh 2 extra options in the if, leave the main command construction same (also no 80 col limit please)

Comment (bcoca) [reply lib/ansible/plugins/connection/ssh.py]: docs should also reflect this needs a newer version of sshpass, most of the installed ones do not allow for the -p option

Comment (thedavidwhiteside) [reply lib/ansible/plugins/connection/ssh.py]: Maybe i am not understanding, but I have to add the extra options (such as PKCS11Provider) after the ssh binary is added to the command. So it would have to be after line 427 right?

I am trying to track down what is the minimum version of sshpass required to support -P.  Should I do an additional check to verify the version of sshpass supports -P and error if it doesn't (and pkcs11_provider was used).  Do you happen to know the files I need to update the documentation for, this is my first contribution to Ansible so i'm still learning my way around.

Comment (bcoca) [reply lib/ansible/plugins/connection/ssh.py]: the docs in the ssh plugin are enough, just add it to the description of the new option

Comment (thedavidwhiteside) [reply lib/ansible/plugins/connection/ssh.py]: I updated the doc to explain the requirement for sshpass 1.6+ and got rid of my 80 col limit, I did format for readability though (example lines 435-438 I formatted so it was easier to read what options were being added).

**Comment by starboarder2001** [comment]
Anything else I need to do for this pull request?

- David

**Comment by starboarder2001** [comment]
@bcoca  Anything else I need to do for this pull request?

David

**Comment by starboarder2001** [comment]
@bcoca I believe I completed all your requested changes.  Just checking in on this one.

**Comment by starboarder2001** [comment]
@bcoca I have rebased a few times, let me know if the changes I made are what you were asking for in your review and I can rebase again.

**Review by thedavidwhiteside — COMMENTED** [review]
I think all the requested changes are complete.

**Comment by thedavidwhiteside** [comment]
@bcoca If there is anything I can do to speed this along let me know.  Thanks!

**Comment by bcoca** [comment]
fix the merge conflict and ping me again (also might want to update to 2.7 as version added)

**Comment by thedavidwhiteside** [comment]
@bcoca Just fixed the merge conflict and updated to version added 2.7 from 2.6

**Review by bcoca — APPROVED** [review]

File: lib/ansible/plugins/connection/ssh.py [hunk 5]
```diff
@@ -506,6 +523,15 @@ def _build_command(self, binary, *other_args):
         # Next, additional arguments based on the configuration.
         #
 
+        # pkcs11 mode allows the use of Smartcards or Ubikey devices
```
Comment (nitzmahone) [inline lib/ansible/plugins/connection/ssh.py]: s/Ubikey/Yubikey/

File: lib/ansible/plugins/connection/ssh.py [hunk 6]
```diff
@@ -489,11 +499,18 @@ def _build_command(self, binary, *other_args):
 
         # If we want to use password authentication, we have to set up a pipe to
         # write the password to sshpass.
-
-        if self._play_context.password:
+        pkcs11_provider = self.get_option("pkcs11_provider")
+        if self._play_context.password or len(pkcs11_provider) >= 1:
```
Comment (nitzmahone) [inline lib/ansible/plugins/connection/ssh.py]: minor style nit, but why `>= 1` instead of `> 0`?

File: lib/ansible/plugins/connection/ssh.py [hunk 7]
```diff
@@ -268,6 +268,16 @@
         vars:
           - name: ansible_ssh_use_tty
             version_added: '2.7'
+        yaml: {key: connection.usetty}
+      pkcs11_provider:
```
Comment (nitzmahone) [inline lib/ansible/plugins/connection/ssh.py:327]: This should probably change in one of two ways to allow PKCS11 support to be enabled on a per-host basis:
a) add `vars: { name=ansible_ssh_pkcs11_provider }` to make the value settable in inventory at the host/group level 
b) add a second config value with vars/ini/env/yaml values like `pkcs11_enabled`

While the actual provider configuration makes sense as a global config/env option (since the provider itself is unlikely to vary by host), there should probably be a separate config+env+var option to turn it on for a given host. Otherwise it's an all-or-none proposition for using it, where in complex environments, you may need both in the same Ansible run. Two options is a bit more hassle, but would allow the global configuration of the provider while hosts could still opt in/out individually (or by group). 

This could also allow you to add a warning that the token PIN is required with PKCS11 (rather than silently failing to auth as it would now by bypassing the PKCS11 path entirely).

File: lib/ansible/plugins/connection/ssh.py [hunk 8]
```diff
@@ -268,6 +268,16 @@
         vars:
           - name: ansible_ssh_use_tty
             version_added: '2.7'
+        yaml: {key: connection.usetty}
```
Comment (nitzmahone) [inline lib/ansible/plugins/connection/ssh.py]: Is this related/necessary for the PKCS11 stuff?

File: lib/ansible/plugins/connection/ssh.py [hunk 9]
```diff
@@ -489,11 +499,18 @@ def _build_command(self, binary, *other_args):
...
         # write the password to sshpass.
-
-        if self._play_context.password:
+        pkcs11_provider = self.get_option("pkcs11_provider")
+        if self._play_context.password or len(pkcs11_provider) >= 1:
             if not self._sshpass_available():
-                raise AnsibleError("to use the 'ssh' connection type with passwords, you must install the sshpass program")
+                raise AnsibleError("to use the 'ssh' connection type with passwords or pkcs11_provider, you must install the sshpass program")
 
+        if self._play_context.password and len(pkcs11_provider) >= 1:
```
Comment (nitzmahone) [inline lib/ansible/plugins/connection/ssh.py]: Related to the discussion of making this inventory-configurable above: seems like using the presence of the password as a silent disable switch could be confusing if someone forgot to explicitly specify a PIN for the host/globally.

File: lib/ansible/plugins/connection/ssh.py [hunk 10]
```diff
@@ -506,6 +523,15 @@ def _build_command(self, binary, *other_args):
         # Next, additional arguments based on the configuration.
         #
 
+        # pkcs11 mode allows the use of Smartcards or Ubikey devices
+        if self._play_context.password and len(pkcs11_provider) >= 1:
```
Comment (nitzmahone) [inline lib/ansible/plugins/connection/ssh.py]: The resolution of the config/var switches discussion above should end up using the same thing here

**Review by nitzmahone — CHANGES_REQUESTED** [review]
This is excellent- I've been wanting to add PIV/cert auth support to Ansible for a long time, so I'm glad to see it. A couple usability and style things.

Also, would be really cool if we could add some basic testing on this via `softhsm` or keyring stuff, but our current connection tests are very generic. I wouldn't necessarily let that block merge, as we have similar needs for WinRM and some others.

Also, I see that you have a fork of paramiko that adds OpenSC support- have you explored getting that merged into upstream paramiko? (*EDIT: just saw your note in the original PR text, so NM, but yeah, would also love to get this plugged in for paramiko once it's available there*)

Thanks!

File: lib/ansible/plugins/connection/ssh.py (hunk 5 above)
Comment (thedavidwhiteside) [reply lib/ansible/plugins/connection/ssh.py]: Fixed. Thanks!

File: lib/ansible/plugins/connection/ssh.py (hunk 6 above)
Comment (thedavidwhiteside) [reply lib/ansible/plugins/connection/ssh.py]: Changed to > 0, I think your right that its cleaner.

File: lib/ansible/plugins/connection/ssh.py (hunk 7 above)
Comment (thedavidwhiteside) [reply lib/ansible/plugins/connection/ssh.py:327]: I went with option a (at least I think I did, as a first time contributor some of the Ansible internals are a little fuzzy).  If you specify the pkcs11 provider then pkcs11 is enabled, if you don't specify the provider then its disabled. I could detect if pkcs11 provider was used but no password was provided, I could print an error that a password/pin is required.  I am trying to avoid adding too many variables, it seems that if you specify a pkcs11_provider then its safe to assume your using pkcs11.  Let me know what you think.

File: lib/ansible/plugins/connection/ssh.py (hunk 8 above)
Comment (thedavidwhiteside) [reply lib/ansible/plugins/connection/ssh.py]: No, I think it snuck in when I merged with the latest develop.  I removed it.

File: lib/ansible/plugins/connection/ssh.py (hunk 9 above)
Comment (thedavidwhiteside) [reply lib/ansible/plugins/connection/ssh.py]: Maybe if pkcs11 provider is set, but no password is set I could print an error message?  Should I just throw an Ansible exception or whats the best method to error on this condition to keep consistent with the rest of the Ansible source code?

File: lib/ansible/plugins/connection/ssh.py (hunk 10 above)
Comment (thedavidwhiteside) [reply lib/ansible/plugins/connection/ssh.py]: Thanks again.  I'll look for each place I check for this condition and update.

File: lib/ansible/plugins/connection/ssh.py (hunk 9 above)
Comment (thedavidwhiteside) [reply lib/ansible/plugins/connection/ssh.py]: Still curious if you think this is a viable solution?

**Comment by starboarder2001** [comment]
@nitzmahone How do the changes look?

**Comment by starboarder2001** [comment]
@nitzmahone How do the changes look?

**Comment by starboarder2001** [comment]
@nitzmahone, I made the changes from your last review.  Do you have any additional requests?

**Comment by starboarder2001** [comment]
@nitzmahone, I made the changes from your last review. Do you have any additional requests?  For those of us that must be compliant with [HSPD12](https://www.dhs.gov/homeland-security-presidential-directive-12), this pull request is really important to us.

**Comment by starboarder2001** [comment]
@nitzmahone, I made the changes from your last review. Do you have any additional requests? For those of us that must be compliant with HSPD12, this pull request is really important to us.

**Comment by carpenike** [comment]
@starboarder2001 -- would this PR also fix issues with Yubikey based SSH logins? I get the following errors when attempting to use my Yubikey, SSH to any host works fine, and you'll see that it's intermittent as sometimes it works fine (h2-0)...

If this is unrelated I'll open up a separate issue.

```
r610-2 | UNREACHABLE! => {
    "changed": false,
    "msg": "Failed to connect to the host via ssh: sign_and_send_pubkey: signing failed: agent refused operation\r\nryan@r610-2: Permission denied (publickey).",
    "unreachable": true
}
r610-1 | UNREACHABLE! => {
    "changed": false,
    "msg": "Failed to connect to the host via ssh: sign_and_send_pubkey: signing failed: agent refused operation\r\nryan@r610-1: Permission denied (publickey).",
    "unreachable": true
}
r610-0 | UNREACHABLE! => {
    "changed": false,
    "msg": "Failed to connect to the host via ssh: ryan@r610-0: Permission denied (publickey).",
    "unreachable": true
}
localhost | SUCCESS => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python"
    },
    "changed": false,
    "ping": "pong"
}
h2-0 | SUCCESS => {
    "ansible_facts": {
        "discovered_interpreter_python": "/usr/bin/python3"
    },
    "changed": false,
    "ping": "pong"
}
```

**Comment by starboarder2001** [comment]
> @starboarder2001 -- would this PR also fix issues with Yubikey based SSH logins? I get the following errors when attempting to use my Yubikey, SSH to any host works fine, and you'll see that it's intermittent as sometimes it works fine (h2-0)...
> 
> If this is unrelated I'll open up a separate issue.
> 
> ```
> r610-2 | UNREACHABLE! => {
>     "changed": false,
>     "msg": "Failed to connect to the host via ssh: sign_and_send_pubkey: signing failed: agent refused operation\r\nryan@r610-2: Permission denied (publickey).",
>     "unreachable": true
> }
> r610-1 | UNREACHABLE! => {
>     "changed": false,
>     "msg": "Failed to connect to the host via ssh: sign_and_send_pubkey: signing failed: agent refused operation\r\nryan@r610-1: Permission denied (publickey).",
>     "unreachable": true
> }
> r610-0 | UNREACHABLE! => {
>     "changed": false,
>     "msg": "Failed to connect to the host via ssh: ryan@r610-0: Permission denied (publickey).",
>     "unreachable": true
> }
> localhost | SUCCESS => {
>     "ansible_facts": {
>         "discovered_interpreter_python": "/usr/bin/python"
>     },
>     "changed": false,
>     "ping": "pong"
> }
> h2-0 | SUCCESS => {
>     "ansible_facts": {
>         "discovered_interpreter_python": "/usr/bin/python3"
>     },
>     "changed": false,
>     "ping": "pong"
> }
> ```

It would simplify using Yubikey, but wouldn't solve your issue.

**Comment by carpenike** [comment]
Thanks. Turns out Ansible doesn't support Yubikey directly and requires the use of ControlMaster + a manual connection to hosts prior to running the playbook.

**Comment by thedavidwhiteside** [comment]
@nitzmahone I synced this with upstream devel.

**Comment by thedavidwhiteside** [comment]
Ready for review @nitzmahone

**Comment by jborean93** [comment]
Just tested this on Fedora 33 with a yubikey 4, can confirm that it works and had no issues.

File: lib/ansible/plugins/connection/ssh.py [hunk 11]
```diff
@@ -578,11 +589,20 @@ def _build_command(self, binary, subsystem, *other_args):
 
         # If we want to use password authentication, we have to set up a pipe to
         # write the password to sshpass.
-
-        if conn_password:
+        pkcs11_provider = self.get_option("pkcs11_provider")
+        if conn_password or len(pkcs11_provider) >= 1:
```
Comment (jborean93) [inline lib/ansible/plugins/connection/ssh.py]: Minor nit: I think we should keep with the normal style here and make the checks simpler. A string is `True` when it's not empty so this is the same.

```suggestion
        if conn_password or pkcs11_provider:
```

**Review by nitzmahone — APPROVED** [review]
LGTM- thanks! 2.11 feature freeze was a couple weeks ago, so we won't be able to merge until 2.11rc1 when we branch off 2.11 and devel becomes 2.12.

File: lib/ansible/plugins/connection/ssh.py [hunk 12]
```diff
@@ -578,11 +589,20 @@ def _build_command(self, binary, subsystem, *other_args):
...
-                raise AnsibleError("to use the 'ssh' connection type with passwords, you must install the sshpass program")
+                raise AnsibleError("to use the 'ssh' connection type with passwords or pkcs11_provider, you must install the sshpass program")
 
+        if conn_password and len(pkcs11_provider) >= 1:
+            self.sshpass_pipe = os.pipe()
+            b_command += [b'sshpass',
+                          b'-d' + to_bytes(self.sshpass_pipe[0], nonstring='simplerepr', errors='surrogate_or_strict'),
+                          b'-P',
+                          b'Enter PIN for ']
+        elif conn_password and len(pkcs11_provider) > 0:
```
Comment (jborean93) [inline lib/ansible/plugins/connection/ssh.py]: I think this check is wrong, it's essentially the same as the if statement above but will never fire because the above statement is always run. Do you mean this?

```suggestion
        elif not conn_password and pkcs11_provider:
```

This check will raise the exception if the password is not set but `pkcs11_provider` is.

File: lib/ansible/plugins/connection/ssh.py [hunk 13]
```diff
@@ -578,11 +589,20 @@ def _build_command(self, binary, subsystem, *other_args):
...
+            raise AnsibleError("To use pkcs11_provider you must specify a password/pin")
+        elif conn_password:
```
Comment (jborean93) [inline lib/ansible/plugins/connection/ssh.py]: There's a bit of duplication of code here. We essentially do the same as the password and pkcs11_provider check but the prompt is optional/based on `sshpass_prompt`.

It seems to be that we should do something like

```python
if conn_password or pkcs11_provider:
    if pkcs11_provider and not conn_password:
        raise AnsibleError('need password/pin ...')

    self.sshpass_pipe = os.pipe()
    b_command += [b'sshpass', b'-d' + to_bytes(self.sshpass_pipe[0], nonstring='simplerepr', errors='surrogate_or_strict')]

    password_prompt = self.get_options('sshpass_prompt')
    if not password_prompt and pkcs11_provider:
        password_prompt = 'Enter PIN for '

    if password_prompt:
        b_command += [b'-P', to_bytes(password_prompt, ...)]

    ...
```

Now the same code is used when setting up sshpass and we now have a way to override the expected prompt even when using PKCS11. The docs for `sshpass_prompt` should be updated to say that it defaults to `Enter PIN for` when `pkcs11_provider` is set.

**Review by jborean93 — CHANGES_REQUESTED** [review]
Just a couple of points, we also need a changelog fragment so we can advertise this new feature in the changelog.

File: lib/ansible/plugins/connection/ssh.py (hunk 13 above)
Comment (thedavidwhiteside) [reply lib/ansible/plugins/connection/ssh.py]: Good catch.  I just took care of the duplication with commit c254a33.

Comment (thedavidwhiteside) [reply lib/ansible/plugins/connection/ssh.py]: I just added a changelog fragment as well.

**Comment by mkrizek** [comment]
!waffling needs_revision

**Comment by samdoran** [comment]
!waffling needs_revision

**Review by jborean93 — APPROVED** [review]
Changes look great, just need to wait until we branch off the 2.11 changes so we can merge this into devel for the 2.12 release.

**Comment by Akasurde** [comment]
@jborean93 Since 2.11 is branched out. Do you want to merge this?

**Comment by jborean93** [comment]
My apologies should have looked at this earlier.

rebuild_merge

**Comment by jborean93** [comment]
Bot is taking it's time, will just merge manually (it passed the rebuild test).

**Comment by Akasurde** [comment]
@thedavidwhiteside Thanks for the contribution and patients. @jborean93 Thanks for the review.
